### DISCUSSION TITLE
What is the best way to integrate web3.js with a React.js front end?

### DISCUSSION THREAD

**Discussion by zerox-alpha** [community/community#139483] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

I'm working on a React.js project that requires interaction with Ethereum smart contracts using web3.js. The project already has the front-end completed, but I'm now integrating the smart contract functions. I have the following questions:

Connecting to MetaMask: How can I reliably connect MetaMask to my React app and handle account changes?
Contract Interaction: What is the best approach to structure web3.js interactions (reading/writing to the contract) in a React project?
Error Handling: How should I manage error handling, especially when a user rejects a transaction or MetaMask connection?
I would appreciate any examples, best practices, or libraries that make this process easier. Thank you!

**Comment by im-shubh** [comment]
Here's how you can handle web3.js integration in your React app effectively:

Connecting to MetaMask:
Use ethereum.request to request access to MetaMask and listen for account or network changes. Here's a sample code snippet for connecting MetaMask:

`useEffect(() => {
  if (window.ethereum) {
    window.ethereum.request({ method: 'eth_requestAccounts' })
      .then(accounts => setAccount(accounts[0]))
      .catch(error => console.error(error));

    window.ethereum.on('accountsChanged', accounts => {
      setAccount(accounts[0]);
    });

    window.ethereum.on('chainChanged', () => {
      window.location.reload();
    });
  } else {
    console.log('MetaMask not detected');
  }
}, []);
`

**Comment by ebndev** [comment]
Hi @zerox-alpha & @im-shubh, thanks for being a part of the GitHub Community!

While we appreciate you wanting to contribute to Discussions, please note: We made the decision to disable the ability to earn Achievements in our Community in order to discourage users from participating in coordinated or inauthentic activity like rapid questions and answers in order to earn badges. You can learn more about this decision in our announcement post here [Achievements will no longer be available in the Community](https://github.com/orgs/community/discussions/106536).

Note that GitHub's [Acceptable Use Policies](https://docs.github.com/en/site-policy/acceptable-use-policies/github-acceptable-use-policies) prohibits coordinated or inauthentic activity like rapid questions and answers. As a result, we'll be unmarking the answer and locking this, and other related posts. 

Any future violations may result in a temporary or indefinite block from the Community. Thanks for understanding.
